src/bin/metadata/mod.rs

```diff
@@ -0,0 +1,433 @@
+        state.methods.iter().for_each(|instr| {
+            instruction_names.push((instr.name.to_string(), instr.name.to_string()))
+        });
+    }
+
+    rename_keywords(&mut instruction_names);
+
+    if let Some(state) = &idl.state {
+        for instr in &state.methods {
+            instruction(&mut f, instr, true, &instruction_names, &ty_names)?;
+        }
+    }
+
+    for instr in &idl.instructions {
+        instruction(&mut f, instr, false, &instruction_names, &ty_names)?;
+    }
+
+    writeln!(f, "}}")?;
+
+    Ok(())
+}
+
+fn instruction(
+    f: &mut File,
+    instr: &IdlInstruction,
+    state: bool,
+    instruction_names: &[(String, String)],
+    ty_names: &[(String, String)],
+) -> std::io::Result<()> {
+    let mut badtys: Vec<String> = instr
+        .args
+        .iter()
+        .filter_map(|field| idltype_to_solidity(&field.ty, ty_names).err())
+        .collect();
+
+    if let Some(ty) = &instr.returns {
+        if let Err(s) = idltype_to_solidity(ty, ty_names) {
+            badtys.push(s);
+        }
+    }
+
+    if badtys.is_empty() {
+        docs(f, 1, &instr.docs)?;
+
+        let name = &instruction_names
+            .iter()
+            .find(|e| *e.0 == instr.name)
+            .unwrap()
+            .1;
+
+        write!(
+            f,
+            "\tfunction {}(",
+            if instr.name == "new" {
+                "initialize"
+            } else {
+                name
+            }
+        )?;
+
+        let mut iter = instr.args.iter();
+        let mut next = iter.next();
+
+        while let Some(e) = next {
+            next = iter.next();
+
+            write!(
+                f,
+                "{} {}{}",
+                idltype_to_solidity(&e.ty, ty_names).unwrap(),
+                e.name,
+                if next.is_some() { "," } else { "" }
+            )?;
+        }
+        let selector = sig(if state { "state" } else { "global" }, &instr.name);
+
+        write!(
+            f,
+            ") selector=hex\"{}\" {}external",
+            hex::encode(selector),
+            if state { "" } else { "view " }
+        )?;
+
+        if let Some(ty) = &instr.returns {
+            writeln!(
+                f,
+                " returns ({});",
+                idltype_to_solidity(ty, ty_names).unwrap()
+            )?;
+        } else {
+            writeln!(f, ";")?;
+        }
+    } else {
+        eprintln!(
+            "instructions {} has arguments of type {} which is not supported on Solidity",
+            instr.name,
+            badtys.join(", ")
+        );
+    }
+
+    Ok(())
+}
+
+fn docs(f: &mut File, indent: usize, docs: &Option<Vec<String>>) -> std::io::Result<()> {
+    if let Some(docs) = docs {
+        for doc in docs {
+            for _ in 0..indent {
+                write!(f, "\t")?;
+            }
+            writeln!(f, "/// {}", doc)?;
+        }
+    }
+
+    Ok(())
+}
+
+fn sig(namespace: &'static str, name: &str) -> Vec<u8> {
```

Comment:
This function can have a better name.

Reply:
I've called it discriminator and added some comments